importer: fix scanning of newlines in enclosed fields for DELIMITED DATA

andyyang890 opened the pull request:
Previously, when using `IMPORT INTO ... DELIMITED DATA`, unescaped
newlines within quoted fields were treated as row terminators instead
of as part of the field. This patch refactors the scanning logic to
correctly interpret these newlines.

Fixes #95906

Release note (bug fix): `IMPORT INTO ... DELIMITED DATA` will now
correctly handle quoted fields that contain unescaped newlines.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97545)
<!-- Reviewable:end -->

andyyang890 committed:
importer: fix scanning of newlines in enclosed fields for DELIMITED DATA

Previously, when using `IMPORT INTO ... DELIMITED DATA`, unescaped
newlines within quoted fields were treated as row terminators instead
of as part of the field. This patch refactors the scanning logic to
correctly interpret these newlines.

Release note (bug fix): `IMPORT INTO ... DELIMITED DATA` will now
correctly handle quoted fields that contain unescaped newlines.

rafiss reviewed (changes requested):
nice, glad this could get fixed! i had some small comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-:-NPOvnTq7UtusrRT5bC6:b-c69kt9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890)

___
*[`pkg/sql/importer/read_import_mysqlout.go` line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOtWgq-zVLPbptT5yn:-NPOtWgq-zVLPbptT5yo:b-dh8lem) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5068fce7b224f017434e9e85b5f472389f5cff1/pkg/sql/importer/read_import_mysqlout.go#L87)):*
> ```Go
> 	row       []rune
> 	err       error
> 	eof       bool
> ```

nice cleanup! this `eof` field really seems needlessly complicated
___
*[`pkg/sql/importer/read_import_mysqlout.go` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOvUGiEFS5osze3siD:-NPOvUGj03QnUg1w1hKi:b-gqm0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L96)):*
> ```Go
> 	var r rune
> 	var w int
> 	var gotEncloser, inEnclosedField, inEscapeSeq, inField bool
> ```

nit: could you add a small descriptive comment for each one of these flags? i think it will make the logic of the for loop easier to understand. specifically, the definitions of "encloser" and the difference between inEnclosedField vs inField would help.
___
*[`pkg/sql/importer/read_import_mysqlout.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOtdKz79Ld9DwosA0h:-NPOtdKz79Ld9DwosA0i:b8rvkra) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L101)):*
> ```Go
> 		r, w, d.err = d.reader.ReadRune()
> 		if d.err == io.EOF {
> 			d.err = nil
> ```

should we add the `if nextLiteral` check here and set `ErrUnexpectedEOF` if true?
___
*[`pkg/sql/importer/read_import_mysqlout.go` line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOuDop4YKC20oSKCOV:-NPOuDop4YKC20oSKCOW:bg0dv5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L141)):*
> ```Go
> 		}
> 
> 		inField = !(r == d.opts.FieldSeparator)
> ```

nit: use `r != d.opts.FieldSeparator`


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-:-NPP7k3Z8Q-uDM02PpOj:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/importer/read_import_mysqlout.go` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOvUGiEFS5osze3siD:-NPP7mNH2hqvzArq_DZ_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L96)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: could you add a small descriptive comment for each one of these flags? i think it will make the logic of the for loop easier to understand. specifically, the definitions of "encloser" and the difference between inEnclosedField vs inField would help.
</blockquote></details>

Done.
___
*[`pkg/sql/importer/read_import_mysqlout.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOtdKz79Ld9DwosA0h:-NPP32uECzDM0N8pftIB:b-e01rz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L101)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we add the `if nextLiteral` check here and set `ErrUnexpectedEOF` if true?
</blockquote></details>

I think that would actually be incorrect since it would disallow ending a file with the escape character. From my reading of the logic before my change, the check here would never actually be reached since it would mean `d.row == nil` (i.e. the row is empty) yet `nextLiteral` is true (i.e. the row had an escape character), which is why I deleted it.
___
*[`pkg/sql/importer/read_import_mysqlout.go` line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOuDop4YKC20oSKCOV:-NPP2kLQ0iCd4D2akCg7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L141)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: use `r != d.opts.FieldSeparator`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
nice, the comments help me understand better now. i just had one request for another test, and should be good to merge after that

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-:-NPU9JYQ5kTItw2AZoWj:btf7bbv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890)

___
*[`pkg/sql/importer/read_import_mysqlout.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOtdKz79Ld9DwosA0h:-NPU8yVd-foX9yJEn3wu:b-a3nef0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L101)):*
> I think that would actually be incorrect since it would disallow ending a file with the escape character

hm, but it seems correct to disallow ending a file with the escape character. at least, no matter what we decide, let's have a test case for that situation.


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97545)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/importer/read_import_mysqlout.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97545#-NPOtdKz79Ld9DwosA0h:-NPUGBdR44Sa_I0PabNu:bsmawow) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec4f7611f24ccc37a5e5b3276e9a795e823facc/pkg/sql/importer/read_import_mysqlout.go#L101)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> I think that would actually be incorrect since it would disallow ending a file with the escape character

hm, but it seems correct to disallow ending a file with the escape character. at least, no matter what we decide, let's have a test case for that situation.
</blockquote></details>

Oh you're right, this is indeed an error case and there's already a test for it here: https://github.com/cockroachdb/cockroach/blob/1c3e46d2f4a3c6977a690409e373e2a7b8b1ca28/pkg/sql/importer/import_stmt_test.go#L478-L487

I believe the check for this is deferred to `FillDatums` to return a more informative error message (i.e. `unmatched literal`).


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for checking! lgtm

andyyang890 commented:
TFTR!

bors r=rafiss
